rfd/0057-automatic-user-provisioning.md

```diff
@@ -0,0 +1,240 @@
+```
+
+Should a user be given `root` access to all commands, they will be
+able to modify any file, including sudoers files.
+
+
+### User deletion
+
+After all of a users sessions are logged out the created user and any
+`sudoers` files that were created for that user will be deleted if
+that user is also a member of the `teleport-system` group.
+
+Users can not be deleted while they have running processes so each
+time a session ends, an attempt to delete the user can happen, if it
+succeeds the sudoers file can also be removed.
+
+If it does not succeed a cleanup process will run periodically that
+will attempt to delete users if they no longer have running processes.
+
+### Multiple matching rols 
+
+Automatic user provisioning will require that all roles matching a
+node via `labels` have `auto_create_user=true`
+
+## UX Examples
+
+### Teleport admin wants each user has a dedicated SSH account defined in their Okta attributes.
+```yaml
+kind: role
+version: v5
+metadata:
+  name: auto-user-groups
+spec:
+  options:
+    # allow auto provisioning of users.
+    auto_create_user: true
+  allow:
+    # username from external okta attribute
+    logins: [ "{{external.username}}" ]
+```
+
+### Teleport admin wants to define which Linux groups each auto-created user will be added to.
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: auto-user-groups
+spec:
+  options:
+    # allow auto provisioning of users.
+    auto_create_user: true
+  allow:
+    # List of each group the user will be added to
+    groups: [ubuntu, docker, ...]
+```
+
+### Teleport admin wants to make each auto-created user a sudoer.
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: users-as-sudoers
+spec:
+  options:
+    # allow auto provisioning of users.
+    auto_create_user: true
+  allow:
+    # add users to the wheel group
+    groups: [wheel]
+    # make it so users in the wheel group will
+    sudoers: ["%wheel ALL=(ALL) NOPASSWD: ALL"]
+```
+
+### Teleport admin wants to define particular commands user will be able to run as root.
+```yaml
+kind: role
+version: v5
+metadata:
+  name: specify-commands-as-sudoers
+spec:
+  options:
+    # allow auto provisioning of users.
+    auto_create_user: true
+  allow:
+    # make it so this specific user can execute `/bin/ `
+    sudoers: ["{{internal.logins}} ALL = (root) NOPASSWD: /usr/bin/systemctl"]
+```
+
+### Teleport admin wants to prohibit some nodes from auto-creating users.
+
+Include the below config for the Teleport node that should not allow automatic user creation:
+
+```yaml
+ssh_service:
+  enabled: "yes"
+  # stops a specific node from auto-creating users
+  disable_auto_create_user: true
+```
+
+Nodes where `diable_auto_create_user` is `false` will still be able to
+have users be automatically created.
+
+### Teleport user has multiple roles but not all of them enable `auto_create_user`
+
+In the situtation where a user has roles as below, the user would not
```

Comment:
This seems overly cautious - this means that all roles assigned to a user must have this on. What is the rationale for that?

Reply:
@r0mant  suggested this, I assumed it was to avoid the possibility that a user accidentally receives too much access

Reply:
I think the idea was to have all roles that _match_ the node you're SSH'ing into require "create_host_user: true", right?

In the example below, both roles match nodes with label `env: example` so most strict setting takes advantage and automatic user creation is basically disabled when you're SSH'ing into such a node.

If, however, your second role did not match `env: example` nodes, then the first role's setting would kick in when you're SSH'ing into `env: example` nodes so auto creation is enabled.

@lxea @klizhentas Does that behavior sound ok?